lib/versions/npmTimer.js

```diff
@@ -0,0 +1,108 @@
+const {PerformanceObserver, performance} = require('perf_hooks');
+const _ = require('underscore');
+require('colors');
+const i18n = require('../i18n/i18n.js');
+i18n.setOrDetectLocale('en');
+
+
+class NpmTimer{
+  constructor(options){
+    this._logger = options.logger;
+    this._packageName = options.packageName;
+    this._version = options.version;
+    this._showSpinner = options.showSpinner || false;
+    this._spinnerStyle = options.spinnerStyle || 'dots';
+    this._interval = options.interval || 750;
+    
+    // define mark and measurement names
+    this._startMark = 'downloadStart' + this._packageName + this._version;
+    this._ongoingMark = 'downloadOngoingMark' + this._packageName + this._version;
+    this._downloadOngoing = 'downloadOngoing' + this._packageName + this._version;
+    this._endMark = 'downloadEnd' + this._packageName + this._version;
+    this._downloadComplete = 'downloadComplete' + this._packageName + this._version;
+    
+    this.observer.observe({entryTypes: ['measure']});
+
+    
+  }
+
+  get observer(){
+    if(typeof this._observer == 'undefined'){
+      this._observer = new PerformanceObserver((items) => {
+        let entry;
+        let strDuration;
+  
+        // find any download ongoing measurements we've made
+        entry = _.last(_.where(items.getEntries(), {name: this._downloadOngoing}));
+        if(entry){
+          // ongoing performance mark 
+          strDuration = __('Downloading and installing {{packageName}} {{version}}... ({{duration}}ms elapsed)', {packageName: this._packageName, version: this._version, duration: entry.duration});
+          if(this._spinner) this._spinner.text = strDuration;
+        }
+        else{
+          // otherwise, find our download complete measurement
+          entry = _.last(_.where(items.getEntries(), {name: this._downloadComplete}));
+          
+          if(entry){
+            strDuration = __('Finished downloading and installing {{packageName}} {{version}} in {{duration}}ms', {packageName: this._packageName, version: this._version, duration: entry.duration});
+            performance.clearMarks();
+            if(this._spinner) this._spinner.succeed(strDuration);
+          }
+        }
+        
+        // log our measurement and make it red if it has taken too long
+        if(!this._showSpinner && entry && strDuration){
+          if(entry.duration > 4000){
+            strDuration = strDuration.red;
+          }
+          this._log(strDuration);
+        }
+        
+      });
+    }
+    return this._observer;
+  }
+
+  _log(msg){
+    if(typeof this._logger.info == 'function'){
+      this._logger.info(msg);
+    }
+    else{ // occurs in subprocess as functions are not serialised/deserialised properly
```

Comment:
Instead of doing that if each time, you can check that in the constructor. 
```
this._logger = options.logger || console;
```
console has all the same functions as logger so it's fine

Reply:
`options.logger` can be defined, but the functions of the logger don't serialise properly when sending the object to a subprocess, so have to check if `logger.info` is a function. Have moved this to the constructor.